sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -133,6 +132,38 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
 }
 
 /**
+  * A pattern that collects the filter and inner joins.
+  *
+  *          Filter
+  *            |
+  *        inner Join
+  *          /    \            ---->      (filters, Seq(plan1, plan2), input)
+  *  inner join    plan2
+  *      /    \
+  *   input    plan1
+  */
+object FilterAndInnerJoins extends PredicateHelper {
+  def unapply(plan: LogicalPlan): Option[(LogicalPlan, Seq[LogicalPlan], Seq[Expression])] =
+    plan match {
+      case f @ Filter(filterCondition, j @ Join(left, right, Inner, None)) =>
+
+        // flatten all inner joins, which are next to each other and has no condition
+        def flattenJoin(plan: LogicalPlan): (LogicalPlan, Seq[LogicalPlan]) = plan match {
+          case Join(left, right, Inner, None) =>
```

Comment:
Why can't these have conditions? Seems this would just go into the returned filters